### PR TITLE
Implement create user endpoint (#18)

### PR DESCRIPTION
**Pull request by jason17h**
* Implement create user endpoint

* Update AWS Service To Add More Calls For Interacting With S3 (#19)

* Modify Upload Function To Return a Promise

* Add Function To Get File From S3

* Fix Return Type On Upload Function

* Add Function To Delete Object In S3

* Move AWS Type Exports To Index

* Add input custom types for CRUD endpoints

* Update user create input types

* Remove create user endpoint implementation from api/user/id

* Implement create user endpoints

* Call create user endpoints on POST request

* Add NextAuth Callbacks and Custom Verification Requests (#24)

* Add callbacks and custom verification request to nextauth

* Added fix to prevent multiple instances of Prisma Client in development due to hot reloading

* Removed unnecessary comments and added function documentation

* Changed comment for mail service function

* Added return types for html and text functins

* Remove whitespace

Co-authored-by: Eric Feng <46411197+feng-eric@users.noreply.github.com>

* Update function documentation

* Implement get program endpoint (#20)

* Start GET endpoint for programs

* Update programs return variable name

* Test new branch name

* Remove prisma include parameters

* Update getProgram

* Fix comments

* Remove empty line

* Store first and last name in user table

* add program sections table, attempt 2 (#23)

* add program sections table, attempt 2

* add on delete cascade, foreign key id. remove email, rename section to class

* Update schema.prisma file

* Update role models

* Add first iteration of usert role functions

* Update teacher and volunteer upsert functions

* Move user update handler to [id].ts

* Update prisma schema

* Revert getUser arg type to string

* Add role enum

* Update services/database/user.ts documentation

Co-authored-by: Eric Feng <46411197+feng-eric@users.noreply.github.com>

* Update services/database/user.ts documentation

Co-authored-by: Eric Feng <46411197+feng-eric@users.noreply.github.com>

* Update services/database/user.ts documentation

Co-authored-by: Eric Feng <46411197+feng-eric@users.noreply.github.com>

* Resolve merge conflicts

* Rename variables

* Link volunteers to users

* Enforce mutual exclusion among user roles

* Update role models

* Import roles from prisma client instead of custom model

* Set role field in updateUser

* Update create user endpoint (#28)

* updated user functions

* Update user.ts

Co-authored-by: Jason Huang <42329517+jason17h@users.noreply.github.com>

Co-authored-by: Mahad Zaryab <43658574+mahadzaryab1@users.noreply.github.com>
Co-authored-by: Eric Feng <46411197+feng-eric@users.noreply.github.com>
Co-authored-by: brandonwong368 <71860803+brandonwong368@users.noreply.github.com>
Co-authored-by: Cindy Wang <h687wang@uwaterloo.ca>
Co-authored-by: sohak22 <76765744+sohak22@users.noreply.github.com>

### Notion ticket link

[Create User Endpoint](https://www.notion.so/uwblueprintexecs/086c1b564dbf478faa649a8880e34b81?v=9d3f8002a0d1477ba395616dfc082af7&p=cbb3f329f3194d91b813950a167e7f4a)

### Implementation description

-

### Steps to test

1.

### What should reviewers focus on?

-

### Checklist

-   [x] My PR name is descriptive and in imperative tense
-   [ ] I have run the linter
-   [ ] I have requested a review from the PL, and/or other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Comment by feng-eric** [comment]
Build seemed to fail btw

**Comment by jason17h** [comment]
> Build seemed to fail btw

Saw that and I think it's because it doesn't know what type the session ID is supposed to be. Is that because it was defined in a custom callback that maybe the linter didn't pick up?

File: pages/api/user/[id].ts [hunk 1]
```diff
@@ -25,8 +26,28 @@ export default async function handle(
...
         ResponseUtil.returnOK(res, user);
         return;
+    } else if (req.method == "PUT") {
+        const session = await getSession({ req });
+        const userId = session ? session.id : req.query.id;
+        const updatedUserData = {
+            id: userId,
+            first_name: req.body.first_name,
+            last_name: req.body.last_name,
+            role: req.body.role,
```
Comment (feng-eric) [inline pages/api/user/[id].ts]: What's `role` vs `role_data`?

File: pages/api/user/[id].ts [hunk 2]
```diff
@@ -25,8 +26,28 @@ export default async function handle(
         }
         ResponseUtil.returnOK(res, user);
         return;
+    } else if (req.method == "PUT") {
+        const session = await getSession({ req });
+        const userId = session ? session.id : req.query.id;
```
Comment (feng-eric) [inline pages/api/user/[id].ts]: Let's also store the user's role in nextauth callback so we can retrieve it from the `session`.

File: services/database/user.ts [hunk 3]
```diff
@@ -35,4 +48,163 @@ async function getUsers() {
...
+    return parent;
+}
+
+/**
+ * Inserts a new programAdmin or updates information corresponding to programAdminData and returns
+ * the newly created or updated programAdmin
+ * @param ParentData - data corresponding to the new programAdmin user
+ * @returns - the newly created or updated programAdmin user
+ */
+
```
Comment (feng-eric) [inline services/database/user.ts]:
```suggestion
```

File: services/database/user.ts [hunk 4]
```diff
@@ -35,4 +48,163 @@ async function getUsers() {
...
+    - create/update the role record
+    - update the user record
+    - return the user, including the role records
+    */
+    const roleData = userInput.role_data;
+    const user = await getUser(userInput.id);
+
+    switch (userInput.role) {
+        case roles.PARENT:
+            if (user.role && user.role !== roles.PARENT) {
```
[the remark this answers is not in the document]
Comment (jason17h) [reply services/database/user.ts:76]: I'm not sure if checking `user.role` outside of the switch statement would make the code cleaner because if `user.role` is not null, then I need a switch statement to check whether the existing `user.role` is the same as the role being passed in, and then I would need another switch statement to actually call the appropriate function - correct me if I'm wrong

File: pages/api/user/[id].ts (hunk 1 above)
Comment (jason17h) [reply pages/api/user/[id].ts]: `role` is the enum name of the role (i.e., `"PARENT"`, `"VOLUNTEER"`, etc.) whereas the `role_data` is the actual data corresponding to that role (e.g., address, phone number, etc.)

File: pages/api/user/[id].ts (hunk 2 above)
Comment (jason17h) [reply pages/api/user/[id].ts]: Should this be in this PR or a separate one?

File: services/database/user.ts [hunk 5]
```diff
@@ -35,4 +48,163 @@ async function getUsers() {
...
+            if (user.role && user.role !== roles.VOLUNTEER) {
+                return;
+            }
+            upsertVolunteer(roleData, userInput.id);
+            break;
+        default:
+            return;
+    }
+
+    const updatedUser = await prisma.user.update({
```
[the remark this answers is not in the document]
Comment (jason17h) [reply services/database/user.ts]: See: https://github.com/uwblueprint/social-diversity-for-children/pull/29/commits/011c1a3f047b3559f35291be1c4fa76403c01b4e

File: pages/api/user/[id].ts (hunk 2 above)
Comment (feng-eric) [reply pages/api/user/[id].ts]: Let's do it in another PR then.

File: services/database/user.ts [hunk 6]
```diff
@@ -35,4 +48,193 @@ async function getUsers() {
...
+            volunteer: true,
+        },
+    };
+
+    let updatedUser;
+
+    switch (userInput.role) {
+        case roles.PARENT: {
+            if (user.role && user.role !== roles.PARENT) {
+                return;
```
Comment (feng-eric) [inline services/database/user.ts:78]: We can do this in another PR but it might be good to return specific error messages to the frontend.

File: services/database/user.ts [hunk 7]
```diff
@@ -35,4 +48,193 @@ async function getUsers() {
...
+                    create: roleData,
+                    update: roleData,
+                    where: { id: user.id },
+                }),
+                prisma.user.update(updateUserArgs),
+            ]);
+            break;
+        }
+        default: {
+            return;
```
Comment (feng-eric) [inline services/database/user.ts:171]: Same comment as above, in a future PR, let's return an error message for this case. Can you add some TODO comments just so we can remember? Thanks!

File: services/database/user.ts [hunk 8]
```diff
@@ -35,4 +48,193 @@ async function getUsers() {
...
+    return updatedUser;
+}
+
+/**
+ * Inserts a new teacher or updates information corresponding to teacherData and returns
+ * the newly created or updated teacher
+ * @param TeacherData - data corresponding to the new teacher user
+ * @returns - the newly created or updated teacher user
+ */
+async function upsertTeacher(
```
Comment (feng-eric) [inline services/database/user.ts]: Is this being used?

File: services/database/user.ts [hunk 9]
```diff
@@ -35,4 +48,193 @@ async function getUsers() {
...
+    return parent;
+}
+
+/**
+ * Inserts a new programAdmin or updates information corresponding to programAdminData and returns
+ * the newly created or updated programAdmin
+ * @param ParentData - data corresponding to the new programAdmin user
+ * @returns - the newly created or updated programAdmin user
+ */
+async function upsertProgramAdmin(
```
Comment (feng-eric) [inline services/database/user.ts]: Is this being used?

**Review by feng-eric — COMMENTED** [review]
Had some comments but other than that, it looked good! Nice work @jason17h @sohak22

File: models/user.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
+import type { roles, province, locale } from "@prisma/client";
+
```
Comment (feng-eric) [inline models/user.ts]: Let's add code documentation for these

**Review by feng-eric — APPROVED** [review]
LGTM but let's get some other people to review it too please!

File: pages/api/user/index.ts [hunk 11]
```diff
@@ -18,10 +21,29 @@ export default async function handle(
...
+            const session = await getSession({ req });
+            const userId = session.id as string;
+            // TODO: add user role to session
+            const updatedUserData = {
+                id: userId,
+                firstName: req.body.firstName,
+                lastName: req.body.lastName,
+                role: req.body.role,
+                roleData: req.body.roleData,
+            } as UserInput;
```
Comment (mahadzaryab1) [inline pages/api/user/index.ts]: instead of typecasting this - you can just define it as a UserInput

`const updatedUserData: UserInput = {...}`

File: pages/api/user/index.ts [hunk 12]
```diff
@@ -18,10 +21,29 @@ export default async function handle(
...
+            const updatedUser = await updateUser(updatedUserData);
```
Comment (mahadzaryab1) [inline pages/api/user/index.ts:38]: are we not doing any validation before updating the user data?

File: services/database/user.ts [hunk 13]
```diff
@@ -35,4 +37,145 @@ async function getUsers() {
...
+        where: { id: parseInt(userInput.id) },
+        include: {
+            teacher: true,
+            parent: true,
+            programAdmin: true,
+            volunteer: true,
+        },
+    };
+
+    let updatedUser;
```
Comment (mahadzaryab1) [inline services/database/user.ts]: add type

**Review by mahadzaryab1 — COMMENTED** [review]
Added a few comments/questions

File: pages/api/user/index.ts (hunk 12 above)
Comment (jason17h) [reply pages/api/user/index.ts:38]: This is happening in a separate PR

Comment (jason17h) [reply pages/api/user/index.ts:38]: See here: https://github.com/uwblueprint/social-diversity-for-children/pull/35

File: models/User.ts [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    province: province;
+    preferredLanguage: locale;
+};
+
+/* Input type for Volunteer inputs */
+export type VolunteerInput = {
+    id: string;
+    phoneNumber?: string;
+    isValid?: boolean;
+    backgroundFormLink?: string;
```
Comment (cindy-wang328) [inline models/User.ts:19]: Curious question, will they have to upload the form at the same time they create their account? Or do they enter the other info first, create the account, and upload the form later? If they upload it later, do we need this background form link field in the input 
Or do we leave the field here but enter null (since it has a ?)😅

File: services/database/user.ts [hunk 15]
```diff
@@ -35,4 +37,145 @@ async function getUsers() {
...
+    /*
+    Flow:
+    - get role and role data from UserInput
+    - create/update the role record
+    - update the user record
+    - return the user, including the role records
+    */
+    const roleData = userInput.roleData;
+    if (!roleData.id) {
+        roleData.id = userInput.id;
```
Comment (cindy-wang328) [inline services/database/user.ts]: Kinda confused, why does userInput and roleData both have id fields that mean the same thing? (i might be fried LOL)

File: models/User.ts (hunk 14 above)
Comment (jason17h) [reply models/User.ts:19]: Good question, I'd assume they create the account first and upload the form later, so yeah I suppose you'd leave this field `null` until then - @feng-eric any comment?

Comment (feng-eric) [reply models/User.ts:19]: Yeah, I agree. It makes more sense to upload the form later.

**Review by victoranyun — APPROVED** [review]

File: models/User.ts (hunk 14 above)
Comment (feng-eric) [reply models/User.ts:19]: After talking to designers, they said it should be optional rather than setting it as null

File: services/database/user.ts (hunk 15 above)
Comment (jason17h) [reply services/database/user.ts]: Good point, these were there because `roleData` previously needed the `id` but it doesn't anymore. I'll get rid of it

File: models/User.ts (hunk 14 above)
Comment (cindy-wang328) [reply models/User.ts:19]: Ok thanks for the clarification!

**Review by cindy-wang328 — APPROVED** [review]
Nice! 👏👏  That's a really long PR, Congrats on getting through all those comments!
